rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+
+These distinct types of tables are a slight fiction: often it will be desirable
+to have some fixed part of runtime or side-loaded tables, e.g. to ensure that
+indexing is reliable. For example, a table representing an array of values
+`[x, y, z]` in the proof might be laid out as
+
+|value1| value2 |
+|------|--------|
+|0     |?runtime|
+|1     |?runtime|
+|2     |?runtime|
+
+where the `value1` entries are fixed in the constraint system. This ensure that
+a malicious prover is not free to create multiple values with the same index,
+such as `[(0,a), (0,x), (1,y), (2,z)]`, otherwise the value `a` might be used
+where the value `x` was intended.
+
+### Combining multiple tables
+
+The current implementation only supports a single table. In order to make
+multiple tables available, we either need to run a separate PLOOKUP argument
+for each table, or to concatenate the tables to form an combined table and
+identify the values from each table in such a way that the prover cannot use
+values from the 'wrong' table.
+
+We already solve a similar problem with 'vector' tables -- tables where a row
+contains multiple values -- where we want to ensure that the prover used an
+entry `(x, y, z)` from the table, and not some parts of different entries, or
+some other value entirely. In order to do this, we generate a randomising field
+element `joint_combiner`, where the randomness depends on the circuit witness,
+and then compute `x + y * joint_combiner + z * joint_combiner^2`. Notice that,
+if the prover wants to select some `(a, b, c)` not in the table and claim that
+it in fact was, they are not able to use the knowledge of the `joint_combiner`
+to choose their `(a, b, c)`, since their choice will change the value of the
+`joint_combiner`.
+
+This is a standard technique for checking multiple equalities at once by
+checking a single equality; to use it, we must simply ensure that the
+randomness depends on all of the (non-constant) values in the equality.
+
+We propose extending each table with an extra column, which we will call the
+`TableID` column, which associates a different number with each table. Since
+the sizes of all tables must be known at constraint system time, we can treat
+this as a fixed column. For example, this converts a lookup `(a, b, c)` in the
+table with ID `1` into a lookup of `(a, b, c, 1)` in the combined table, and
+ensures that any value `(x, y, z)` from table ID `2` will not match that
+lookup, since its representative is (x, y, z, 2)`.
+
+To avoid any interaction between the `TableID` column and any other data in
+the tables, we use `TableID * joint_combiner^max_joint_size` to mix in the
+`TableID`, where `max_joint_size` is the maximum length of any lookup or table
+value. This also simplifies the 'joining' of the `table_id` polynomial
+commitment, since we can straightforwardly scale it by the constant
+`joint_combiner^max_joint_size` to find its contribution to every row in the
+combined table.
```

Comment:
I'm always scared of off-by-one's. What about scaling it with `joint_combiner^0 = 1`?

So the previous example would look like `2 + x * joint_combiner + y * joint_combiner^2 +
x * joint_combiner^3`

Reply:
The idea behind not doing this was that we can avoid paying the cost of computing that `joint_combiner` highest power if we only have a single table with ID 0.